### PR TITLE
Added dynamic reconfigure, auto white balance

### PR DESCRIPTION
**Pull request by lwan-zz**
Exposed current camera settings to dynamic reconfigure: 
- Gain (and auto gain toggle)
- Brightness (and auto brightness toggle -- neither seem to do anything to the image)
- Exposure (and auto exposure)
- Gamma
- Auto white balance enable

Other settings will need additional hooks into the Pylon API (custom white balance setttings, trigger settings, etc.)

Also set default MTU from 3000 to 1500 in case it's not set in the .launch file.

### CODE REVIEW

File: package.xml [hunk 1]
```diff
@@ -47,8 +47,9 @@
   <run_depend>roscpp</run_depend>
   <run_depend>roslaunch</run_depend>
   <run_depend>sensor_msgs</run_depend>
-  <!--run_depend>std_srvs</run_depend-->
   <run_depend>nodelet</run_depend>
+  <run_depend>dynamic_reconfigure</run_depend>
+  <run_depend>roslint</run_depend>
```
Comment (jonbinney) [inline package.xml]: Why did you make roslint a run_depend?

File: src/pylon_camera/main.cpp [hunk 2]
```diff
@@ -35,19 +35,34 @@
...
 int main(int argc, char **argv)
 {
     ros::init(argc, argv, "pylon_camera_node");
+
+
     ros::NodeHandle nh_private("~");
     ros::NodeHandle nh_image;
 
+	dynamic_reconfigure::Server<pylon_camera::CameraSettingsConfig> server;
+    dynamic_reconfigure::Server<pylon_camera::CameraSettingsConfig>::CallbackType f;
```
Comment (jonbinney) [inline src/pylon_camera/main.cpp]: This shouldn't be in `main()` now, right? You moved it to the PylonCameraNode class.

File: src/pylon_camera/main.cpp [hunk 3]
```diff
@@ -35,19 +35,34 @@
...
+    //f = boost::bind(&pylon_camera::PylonCameraNode::reconfigureCallback, _1, _2);
+    f = boost::bind(&callback, _1, _2);
+	server.setCallback(f);
+
     pylon_camera::PylonCameraNode pylon_camera_node(nh_private, nh_image);
 
     ROS_INFO_STREAM("Start image grabbing if node connects to topic with "
         << "a frame_rate of: " << pylon_camera_node.frameRate() << " Hz");
-
+
```
Comment (jonbinney) [inline src/pylon_camera/main.cpp]: Remove white space. Actually `main.cpp` should probably not have any changes at all in this PR.

File: src/pylon_camera/pylon_camera_node.cpp [hunk 4]
```diff
@@ -34,6 +34,7 @@
 #include <vector>
 #include "boost/multi_array.hpp"
 
+
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_node.cpp]: Try not to make whitespace changes in a feature PR unless there is a strong reason to. They make it harder to find the real changes in the diff.

File: src/pylon_camera/pylon_camera_node.cpp [hunk 5]
```diff
@@ -178,14 +222,13 @@ bool PylonCameraNode::startGrabbing()
...
-        set_user_output_srvs_.at(i) = nh_private_->advertiseService< camera_control_msgs::SetBool::Request,
-                                                            camera_control_msgs::SetBool::Response >(
-                                            srv_name,
-                                            boost::bind(&PylonCameraNode::setUserOutputCB,
-                                                        this,
-                                                        i,
-                                                        _1,
-                                                        _2));
+        set_user_output_srvs_.at(i) = nh_private_->advertiseService< camera_control_msgs::SetBool::Request, 
+                                                                     camera_control_msgs::SetBool::Response >(
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_node.cpp:200]: Is this just a whitespace change?

File: src/pylon_camera/pylon_camera_parameter.cpp [hunk 6]
```diff
@@ -204,6 +207,16 @@ void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
             }
         }
     }
+
+
+    
+    if ( nh.hasParam("whitebalance_auto") )
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_parameter.cpp:197]: Dynamic reconfigure should call your callback with the values set as ROS params at startup - is there a reason you also check the parameter manually here?

File: src/pylon_camera/pylon_camera_parameter.cpp [hunk 7]
```diff
@@ -71,6 +72,8 @@ PylonCameraParameter::~PylonCameraParameter()
 
 void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
 {
+    ROS_INFO_STREAM("Reading parameters from paramter server");
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_parameter.cpp]: Spelling: `paramter`

File: src/pylon_camera/pylon_camera_parameter.cpp [hunk 8]
```diff
@@ -45,7 +45,7 @@ PylonCameraParameter::PylonCameraParameter() :
         binning_y_given_(false),
         downsampling_factor_exp_search_(1),
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_parameter.cpp:46]: Can we get rid of this `PylonCameraParameter` class completely and replace it with the one that dynamic-reconfigure generates automatically? That will require adding the remaining parameters to dynamic_reconfigure (e.g. `mtu_size`), but that will be a good idea anyway.

File: src/pylon_camera/pylon_camera_node.cpp (hunk 5 above)
Comment (jonbinney) [reply src/pylon_camera/pylon_camera_node.cpp:200]: Actually, screw backwards compatibility in this case. Can you just get rid of the individual parameter setting services? They are obsolete once we can change things with dynamic_reconfigure.

File: src/pylon_camera/pylon_camera_parameter.cpp [hunk 9]
```diff
@@ -204,6 +207,16 @@ void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
             }
         }
     }
+
+
+    
+    if ( nh.hasParam("whitebalance_auto") )
+    {
+        nh.getParam("whitebalance_auto", whitebalance_auto_);
+        std::cout << "continuous auto white balance set" << std::endl;
```
Comment (jonbinney) [inline src/pylon_camera/pylon_camera_parameter.cpp]: Don't use `cout` here. If you need to output, use `LOG_INFO`.

**Review by jonbinney — CHANGES_REQUESTED** [review]

File: package.xml (hunk 1 above)
Comment (lwan-zz) [reply package.xml]: Bad habit to make build_depends also run_depends. Will take out.

File: src/pylon_camera/main.cpp (hunk 2 above)
Comment (lwan-zz) [reply src/pylon_camera/main.cpp]: Yep. Taking it out.

File: src/pylon_camera/main.cpp (hunk 3 above)
Comment (lwan-zz) [reply src/pylon_camera/main.cpp]: Will do. Reverting changes to `main.cpp`

File: src/pylon_camera/pylon_camera_node.cpp (hunk 4 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_node.cpp]: Will remove.

File: src/pylon_camera/pylon_camera_node.cpp (hunk 5 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_node.cpp:200]: Yep. I'll remove the services per our conversation today.

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 8 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_parameter.cpp:46]: Yep, this should be removed as part of moving everything to dynamic reconfigure.

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 7 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_parameter.cpp]: Okey doke. Will change.

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 6 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_parameter.cpp:197]: @jonbinney:  
Wouldn't I want to see if the parameter is already set? Dynamic reconfigure should be setting it at startup, but what if the launchfile sets a different value?

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 9 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_parameter.cpp]: Will do.

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 6 above)
Comment (jonbinney) [reply src/pylon_camera/pylon_camera_parameter.cpp:197]: Dynamic reconfigure actually does this. It checks for a ROS parameter with the name of the parameter at startup. So if you set it in the launchfile, dynamic reconfigure will pass that value to your callback.

File: src/pylon_camera/pylon_camera_parameter.cpp (hunk 9 above)
Comment (lwan-zz) [reply src/pylon_camera/pylon_camera_parameter.cpp]: Any reason not to use `ROS_INFO_STREAM` ?

Comment (jonbinney) [reply src/pylon_camera/pylon_camera_parameter.cpp]: ROS_INFO_STREAM is fine too - as long as the output goes through the ROS logging system.

File: package.xml (hunk 1 above)
Comment (lwan-zz) [reply package.xml]: @jonbinney doesn't look like it'll build without roslint as a run_depend:

`CMake Error at /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:224 (message):
  catkin_package() DEPENDS on the catkin package 'roslint' which must
  therefore be listed as a run dependency in the package.xml
Call Stack (most recent call first):
  /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:102 (_catkin_package)
  pylon_camera/CMakeLists.txt:40 (catkin_package)
`

**Comment by jonbinney** [comment]
Looks good!

**Pull request merged**